### PR TITLE
Support loading kernel, device tree, trust cache from im4p files

### PR DESCRIPTION
**Pull request by qmfrederik**
This adds support for extracting the kernel, device tree and trust cache from im4p files.

It turns out qemu doesn't link with OpenSSL after all, but does use GnuTLS which uses libtasn1, which supports parsing ASN.1 files as well.

It does require you to define a schema for the ASN.1 structure which you're parsing (hence the `img4.asn1` file), which is then converted into a "definitions array" using `asn1Parser -o img4.c -n img4_definitions_array img4.asn1`.

The code assumes:
- The file is an IM4P file if `asn1_der_decoding` can successfully parse the file.
- The payload is LZFSE-compressed if it starts with the string `bvx`.

### CODE REVIEW

**Comment by TrungNguyen1909** [comment]
trust caches use "rtsc" as payload type AFAIK, can you verify if it works or not?

**Comment by TrungNguyen1909** [comment]
Also, you wouldnt want qemu exits if im4p parser failed because there are cases where you want qemu to load the unpacked data as it.

For instance, ramdisk: its filesystem need to be decompressed. In that case, you wouldnt want to repack it into im4p format, right?

**Comment by qmfrederik** [comment]
@TrungNguyen1909 , correct, let me rephrase that a bit:

The code assumes:

- The file is an IM4P file if `asn1_der_decoding` can successfully parse the file. If it is not an IM4P file, the file is loaded as is.
- The payload is LZFSE-compressed if it starts with the string `bvx`. If the payload does nto start with `bvx`, the payload is loaded as is.

**Comment by qmfrederik** [comment]
@TrungNguyen1909 I added support for the trustcache and a couple of sanity checks (to make sure the data which is being loaded is a valid MACH object or a valid trust cache).

I didn't implement the ram disk since it needs to be decompressed anyway.

**Comment by TrungNguyen1909** [comment]
Errors in `extract_im4p_payload` should be returned to the caller instead of `exit(EXIT_FAILURE)`.

And when it fails, we should fallback to load the file as it.

**Comment by qmfrederik** [comment]
> Errors in extract_im4p_payload should be returned to the caller instead of exit(EXIT_FAILURE).
> And when it fails, we should fallback to load the file as it.

Thanks, in that case, I'd suggest to change the semantics of `extract_im4p_payload` to be closer to those of `g_file_get_contents`. That would mean returning `TRUE`/`FALSE` depending on whether the operation succeeds.

If the file is not an DER-encoded ASN.1 file which conforms to the IM4P payload specification, the raw content is always returned (as is the case in the current implementation).

But I'm not sure that falling back to the raw payload is _always_ the right thing to do.

For example, when `asn1_der_decoding` succeeds, the file **is** a DER-encoded ASN.1 file which conforms to the IM4P payload specification. This means the raw content can _never_ be a valid kernel, trust cache, device tree or ram disk file. Returning the raw content will only result in a failure later down, right?

In that case, I'd suggest to print an error message (for example: "expeting a krnl file but we got a dtre file") and just return `FALSE`, so that the calling code can clean up and exit.

What about we change the semantics like this:

- Change `extract_im4p_payload` to return a `gboolean` just like `g_file_get_contents` does, and have it return `FALSE` if `g_file_get_contents` fails.
- Failures in `asn1_array2tree` and `asn1_create_element` are some corner cases, since this code is not operating on user input and not supposed to fail. If they fail, you probably hit an out of memory situation or something similar. So we could probably just `assert` instead`.
- If `asn1_der_decoding` fails, the file is returned "as is" (this is already the case)
- If `asn1_der_decoding` succeeds, but any of the ASN.1-manipulating functions fail, the file is very probably invalid (e.g. the wrong type, wrong magic,...). So call `error_report` but return `FALSE` instead of calling `exit(EXIT_FAILURE)`?

Though in the end, all calls to `extract_im4p_payload` will probably just abort if `extract_im4p_payload` returns `FALSE` (just like they abort now if `g_file_get_contents` returns false)?

**Comment by TrungNguyen1909** [comment]
Ah, I got it.

Sorry I missed that earlier.

Thanks.

File: hw/arm/xnu.c [hunk 1]
```diff
@@ -150,17 +152,127 @@ static void macho_dtb_node_process(DTBNode *node)
...
+        }
+
+        len = 4;
+        if ((ret = asn1_read_value(img4, "type", type, &len)) != ASN1_SUCCESS) {
+            error_report("Failed to read the im4p type in file '%s': %d.", filename, ret);
+            exit(EXIT_FAILURE);
+        }
+
+        if (strncmp(type, payload_type, 4) != 0) {
+            error_report("Could parse ASN.1 data in file '%s' because it is not a %s object.", filename, payload_type);
```
Comment (TrungNguyen1909) [inline hw/arm/xnu.c:207]:
```suggestion
            error_report("Could parse ASN.1 data in file '%s' because it is not a %s object, expected %s object.", filename, payload_type, type);
```
